### DISCUSSION TITLE
Collect Data and Statistics

### DISCUSSION THREAD

**Discussion by muhammadelmogy** [grblHAL/core#292] [Ideas]
I want to be able to collect some data while the machine is running for example:

- Total time the machine run
- Total completed jobs
- Total distance the machine moved for all axes
- and more

Those data could be useful for example in the case of the total time the machine run and/or the total distance the the axes has moved could be used for automating a lubrication system for the linear rails and other things that need lubrication.

The thing is I do not know which is more suitable to do this job, is it the sender or the controller?

**Comment by terjeio** [comment]
> The thing is I do not know which is more suitable to do this job, is it the sender or the controller?

The [controller](https://github.com/grblHAL/Plugin_odometer/). How to count completed jobs? #of M30 commands?

**Reply by muhammadelmogy** [reply]
This idea was originally because I wanted to automate the lubrication of the rails and may be other jobs so the mentioned plug-in is more than enough for this job, but I am not aware of a sender that can handle such a thing, Does ioSender capable of processing the plug-in data and make some decision based on those data?

>  How to count completed jobs? #of M30 commands?

I am not sure really, does all G-Code programs always end with M30?

**Reply by terjeio** [reply]
> Does ioSender capable of processing the plug-in data and make some decision based on those data?

All senders are capable of handling such data if code is written for it. Currently ioSender, and likely all other senders, do not have code to handle it - so for now you have to manually request the information and act upon it. Or write an app to do it and run it periodically.

Lubrication of linear rails can be in the 20-1000km range depending on type and load (source: HIWIN lubrication instructions) if I read the graphs correctly, for a hobby machine this means not very often? An axis running 2m/min 24/7 would require lubrication about once a week for the lower end (high load 20km interval)? The chinese HIWIN knock-offs might be a different story - but I guess they do not come with lubrication specs.

> I am not sure really, does all G-Code programs always end with M30?

They should but does not always do so... So IMO not easy to count jobs in a reliable way.
